core/stdlib/std.ncl

```diff
@@ -2800,6 +2800,250 @@
+      "%
+        = std.contract.from_predicate is_semver_prefix,
+      SemverReq
+        | doc m%"
+        A contract for semantic version ("semver") requirements.
+
+        Nickel supports two kinds of requirements: semver-compatible
+        requirements and exact version requirements. Semver-compatible
+        requirements take the form "major.minor.patch", where minor and patch
+        are optional. Their semantics are:
+
+        - "1.2.3" will match all versions having major version 1, minor version 2,
+          and patch version at least 3.
+        - "1.2" will match all versions having major version 1 and minor version
+          at least 2.
+        - "1" will match all versions having major version 1.
+        - a semver-compatible requirement will never match a prerelease version.
+
+        Exact version requirements take the form "=major.minor.patch-pre", where
+        the prerelease tag is optional, but major, minor, and patch are all required.
+
+        # Examples
+
+        ```nickel multiline
+        "1.2" | SemverReq
+        # => "1.2"
+
+        "=1.2" | SemverReq
+        # => error: contract broken by a value
+
+        "1.2.0" | SemverReq
+        # => "1.2.0"
+
+        "=1.2.0" | SemverReq
+        # => "=1.2.0"
+
+        "1.2.0-pre1" | SemverReq
+        # => error: contract broken by a value
+
+        "=1.2.0-pre1" | SemverReq
+        # => "=1.2.0-pre1"
+        ```
+      "%
+        = std.contract.from_predicate is_semver_req,
+      # TODO: bikeshedding opportunity: which fields should be optional?
+      Manifest = {
+        name
+          | String
+          | doc m%"
+          The name of this package.
+          "%,
+
+        version
+          | String
+          | Semver
+          | doc m%"
+          The version of this package.
+
+          Any semantic version is accepted, but the build metadata field has no effect when matching versions.
+          "%,
+
+        nickel_version
+          | String
+          | SemverPrefix
+          | doc m%"
+          The minimal nickel version required for this package.
+          "%,
+
+        authors
+          | Array String
+          | doc m%"
+          The authors of this package.
+          "%,
+
+        description
+          | String
+          | doc m%"
+          A description of this package.
+          "%,
+
+        keywords
+          | Array String
+          | optional
+          | doc m%"
+          A list of keywords to help people find this package.
+          "%,
+
+        # TODO: maybe restrict this to be a valid SPDX 2.3 license expression?
+        # Cargo allows anything here, but applies restrictions when trying to
+        # publish to crates.io.
+        license
+          | String
+          | optional
+          | doc m%"
+          The name of the license that this package is available under.
+          "%,
+
+        dependencies
+          | {
+            _ : [|
+              'Path String,
+              'Git {
```

Comment:
Honestly, it was just because the first version had less fields, and then I didn't split it out as the number of fields grew. I'm not sure that static typing is a big concern here, as we don't have functions that consume and produce `Manifest`s. (At least, not yet. But by the time we do, maybe we'll also have `let type`...)

Reply:
Well, if we put it in the stdib, people might want to do that, so we could be tempted to leave it as it is and then later split it in a nicer way once we have let types, without disturbing user code.

On the other hand if we split it into contracts now this reduces what users can do with manifests but we can still introduce let-types later and that would be a backward-compatible improvement, which has been our strategy for many things up to now. So indeed if the stdlib doesn't really need the statically typed side, I guess we can go both ways.